### PR TITLE
Splitting the heavy sources in parallel

### PR DESCRIPTION
**Pull request by micheles**
In the current master we split the sources by using a single core in the controller node. With this pull requests we split the heavy sources in the workers by producing a task per heavy source.  The improvement in the case of a reduced SHARE model is the following:

```
managing sources               971 seconds (master)
managing sources               596 seconds (this branch)
```

This is quite good and has the advantage that the management of heavy sources and light sources is completely decoupled, so that further optimizations could be devised in the future.

The demos are running here: https://ci.openquake.org/job/zdevel_oq-engine/2150/

### CODE REVIEW

File: openquake/commonlib/parallel.py [hunk 1]
```diff
@@ -282,17 +282,21 @@ def starmap(cls, task, task_args, name=None):
...
+                cls.progress('Submitting %s "%s" tasks', nargs, self.name)
             if isinstance(a[-1], Monitor):  # add incremental task number
                 a[-1].task_no = i
             self.submit(*a)
         return self
 
     @classmethod
     def apply(cls, task, task_args,
-              concurrent_tasks=executor._max_workers,
+              concurrent_tasks=executor._max_workers * 2,
```
Comment (ptormene) [inline openquake/commonlib/parallel.py]: Perhaps we could add a comment about this change

File: openquake/commonlib/source.py [hunk 2]
```diff
@@ -728,6 +727,28 @@ def source_info_iadd(self, other):
 ])
 
 
+def split_filter(src, sites, max_dist, random_seed):
+    """
+    :param src: an heavy source
+    :param sites: the sites affected by the source
+    :param max_dist: maximum distance for the current TRT
+    :random_seed: used only for event based calculations
+    """
```
Comment (ptormene) [inline openquake/commonlib/source.py]: Please add the description of the return value

File: openquake/commonlib/source.py [hunk 3]
```diff
@@ -755,109 +775,104 @@ def __init__(self, csm, maximum_distance, concurrent_tasks,
...
                 start += nr
 
-    def get_sources(self, kind, tile):
+    def gen_blocks(self, srcs_times, tiles):
         """
-        :param kind: a string 'light', 'heavy' or 'all'
-        :param tile: a :class:`openquake.hazardlib.site.Tile` instance
-        :returns: the sources of the given kind affecting the given tile
+        :param srcs_times:
+            a 5-uple (src, sites, split_sources, filter_time, split_time)
```
Comment (ptormene) [inline openquake/commonlib/source.py]: Please describe also tiles and what the generator yields

File: openquake/commonlib/source.py (hunk 2 above)
Comment (micheles) [inline openquake/commonlib/source.py]: Done

File: openquake/commonlib/parallel.py (hunk 1 above)
Comment (micheles) [inline openquake/commonlib/parallel.py]: Using a talking variable name

File: openquake/commonlib/source.py [hunk 4]
```diff
@@ -728,6 +727,30 @@ def source_info_iadd(self, other):
 ])
 
 
+def split_filter(src, sites, max_dist, random_seed):
+    """
+    :param src: an heavy source
+    :param sites: the sites affected by the source
+    :param max_dist: maximum distance for the current TRT
+    :random_seed: used only for event based calculations
+    ;returns:
```
Comment (ptormene) [inline openquake/commonlib/source.py]: Typo: please change `;` into `:`

**Comment by ptormene** [comment]
If also demos are green, LGTM

**Pull request merged**